File: packages/db/src/entities/Category.ts
import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  CreateDateColumn,
  UpdateDateColumn,
  ManyToOne,
  OneToMany,
  Index,
  JoinColumn,
} from 'typeorm';

@Entity('categories')
@Index(['disciplineId', 'slug'], { unique: true })
export class Category {
  @PrimaryGeneratedColumn('increment')
  id!: number;

  @Index()
  @Column({ type: 'bigint', nullable: false })
  disciplineId!: number;

  @Column({ type: 'varchar', length: 255, nullable: false })
  name!: string;

  @Column({ type: 'varchar', length: 255, nullable: false })
  slug!: string;

  @Index()
  @Column({ type: 'bigint', nullable: true })
  parentId!: number | null;

  @Column({ type: 'text', nullable: true })
  description!: string | null;

  @Column({ type: 'int', default: 0 })
  ord!: number;

  @Index()
  @Column({ type: 'bigint', nullable: true })
  createdBy!: number | null;

  @CreateDateColumn()
  createdAt!: Date;

  @UpdateDateColumn()
  updatedAt!: Date;

  // Self-referential relationship
  @ManyToOne(() => Category, (category) => category.children, { nullable: true, onDelete: 'CASCADE' })
  @JoinColumn({ name: 'parentId' })
  parent?: Category | null;

  @OneToMany(() => Category, (category) => category.parent)
  children?: Category[];
}

File: backend/internal/model/category.go
package model

import "time"

type Category struct {
	ID           string     `json:"id" firestore:"-"`
	DisciplineID string     `json:"disciplineId" firestore:"disciplineId"`
	Name         string     `json:"name" firestore:"name"`
	Slug         string     `json:"slug" firestore:"slug"`
	Description  string     `json:"description" firestore:"description"`
	ParentID     *string    `json:"parentId" firestore:"parentId"`
	OwnerUID     string     `json:"ownerUid" firestore:"ownerUid"`
	CreatedAt    time.Time  `json:"createdAt" firestore:"createdAt"`
	UpdatedAt    time.Time  `json:"updatedAt" firestore:"updatedAt"`
	Children     []Category `json:"children,omitempty" firestore:"-"`
}

type CreateCategoryRequest struct {
	Name        string  `json:"name"`
	Description string  `json:"description"`
	ParentID    *string `json:"parentId"`
}

type UpdateCategoryRequest struct {
	Name        *string `json:"name"`
	Description *string `json:"description"`
	ParentID    *string `json:"parentId"`
}
